test/model/src/core/core.cpp

```diff
@@ -0,0 +1,138 @@
+ * means.
+ *
+ * In jurisdictions that recognize copyright laws, the author or authors
+ * of this software dedicate any and all copyright interest in the
+ * software to the public domain. We make this dedication for the benefit
+ * of the public at large and to the detriment of our heirs and
+ * successors. We intend this dedication to be an overt act of
+ * relinquishment in perpetuity of all present and future rights to this
+ * software under copyright law.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND,
+ * EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF
+ * MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT.
+ * IN NO EVENT SHALL THE AUTHORS BE LIABLE FOR ANY CLAIM, DAMAGES OR
+ * OTHER LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE,
+ * ARISING FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR
+ * OTHER DEALINGS IN THE SOFTWARE.
+ */
+
+#include <algorithm>
+#include <iostream>
+
+#include "model/core.h"
+#include "wiredtiger.h"
+
+namespace model {
+
+/*
+ * NONE --
+ *     The "None" value.
+ */
+const data_value NONE = data_value::create_none();
+
+/*
+ * kv_item::~kv_item --
+ *     Delete the instance.
+ */
+kv_item::~kv_item()
+{
+    for (auto i = _updates.begin(); i != _updates.end(); i++)
+        delete *i;
+}
+
+/*
+ * kv_item::add_update --
+ *     Add an update.
+ */
+int
+kv_item::add_update(kv_update &&update, bool must_exist, bool must_not_exist)
+{
+    std::lock_guard lock_guard(_lock);
+    kv_update::timestamp_comparator cmp;
+
+    /* If this is a non-timestamped update, there cannot be existing timestamped updates. */
+    if (update.global()) {
+        if (!_updates.empty() && !_updates[_updates.size() - 1]->global())
+            return (EINVAL);
+    }
+
+    /* Position the update. */
+    auto i = std::upper_bound(_updates.begin(), _updates.end(), &update, cmp);
+
+    /* If need be, fail if the key does not exist. */
+    if (must_exist) {
+        if (_updates.empty())
+            return (WT_NOTFOUND);
+
+        auto j = i;
+        if (j == _updates.begin() || (*(--j))->value() == NONE)
+            return (WT_NOTFOUND);
+    }
+
+    /* If need be, fail if the key exists. */
+    if (must_not_exist && !_updates.empty()) {
+        auto j = i;
+        if (j != _updates.begin() && (*(--j))->value() != NONE)
+            return (WT_DUPLICATE_KEY);
+    }
+
+    /* Insert. */
+    _updates.insert(i, new kv_update(std::move(update)));
+    return (0);
+}
+
+/*
+ * kv_item::contains_any --
+ *     Check whether the table contains the given value. If there are multiple value associated with
+ *     the given timestamp, return true if any of them match.
+ */
+bool
+kv_item::contains_any(const data_value &value, uint64_t timestamp)
+{
+    std::lock_guard lock_guard(_lock);
+    kv_update::timestamp_comparator cmp;
+
+    auto i = std::upper_bound(_updates.begin(), _updates.end(), timestamp, cmp);
```

Comment:
Could `std::equal_range` be used?

Reply:
I guess I misunderstand. I was thinking that if the timestamp comparator defined equality as <= timestamp, something like:
```
auto cmp = [&timestamp](const kv_update& update){ return udpate.timestamp <= timestamp; }
```
You would then have an iterator pair you can traverse that contains all the updates you are interested in.

Reply:
That's fair, thank you for explaining! At this point, I think it makes sense for this function to stay as it is—provided that it is understandable given the new comments.

Reply:
As you feel is best.